### ISSUE TITLE
Consul-template fails to evaluate specific syntax with consul 1.3.0+

### ISSUE THREAD

**Issue by bhn26** [hashicorp/consul-template#1156]
### Consul Template version

consul-template v0.19.5

### Consul version

Consul v1.3.0

### Configuration

```hcl
consul-template.conf:
consul {
  address = "127.0.0.1:8500"
}

syslog {
  enabled = false

  // This is the name of the syslog facility to log to.
  facility = "LOCAL5"
}

db.conf:
template {
  source = "/etc/consul-template/db.sh.ctmpl"

  destination = "/etc/consul-template/db.sh"

  command = "bash /etc/consul-template/db.sh"

  command_timeout = "90s"

  perms = 0755

  backup = false
}
```

```liquid
{{ range service "unique_tag.example_service@dc1" }}
  echo "{{ .Node }}"
{{ end }}
```

### Command

```shell
/usr/bin/consul-template -config /etc/consul-template/conf
```

### Debug output

I've provided the logs for consul 1.30 and 1.2.3 in this gist:
https://gist.github.com/bhn26/30faf646e195cd84e5878a040bd3b9fb

### Expected behavior

Expected behavior is that the following consul-template syntax:

```
{{ service "<TAG>.<NAME>@<DATACENTER>~<NEAR>|<FILTER>" }}
```

From the used template above, would render a db.sh file containing something like the following:
```
# cat db.sh

  echo "test-client"

```

### Actual behavior

Using consul versions 1.3.0, 1.3.1, and 1.4.0, we the db.sh file doesn't contain anything and from the logs, it seems like it doesn't pick up the service correctly via that <TAG> syntax.

### Steps to reproduce

0. Set up a consul cluster with tags, nodes, and services of similar name as defined in the template.
1. Install consul 1.3.0+, install consul-template 0.19.5
2. Set up the conf/db.conf and db.sh.ctmpl as defined above.
3. Run the consul-template command and see that the rendered file doesn't contain any information.
4. Switch to consul 1.2.3 and try to run the command again and information will be present in the db.sh file.

**Comment by pierresouchay** [comment]
@bhn26 Probably fixed by https://github.com/hashicorp/consul/pull/4944 (broken Consul compatibility)...

@Aestek @banks Do you think it is related?

**Comment by banks** [comment]
I don't think this has been fixed by that PR @pierresouchay - OP states it's broken with 1.3.1 and 1.4.0.

I think it might be the same root-cause as https://github.com/hashicorp/consul/pull/4916 - the logs show CT gets the result just fine but doesn't consider it to be changed which may be related to IsSame or similar not being updated correctly to work with the internal change.

The puzzle is that the external API should not have changed in Consul even if CT is still using the older API client code etc. We'll investigate but this should probably be a bug on consul if we broke backwards compat in this way.

**Comment by banks** [comment]
FWIW I just tried this locally, almost verbatim and it works for me with consul 1.4.0 and latest consul-tempalte build.

I registered the service instance locally with:

```hcl
service {
  name = "example_service"
  port = 9999
  tags = ["unique_tag"]
}
```

In consul.hcl and started dev agent: `consul agent -dev -config-file consul.hcl`

Then used the same template config as posted here:

db.conf
```hcl
template {
  source = "./db.sh.ctmpl"

  destination = "./db.sh"

  command = "bash ./db.sh"

  command_timeout = "90s"

  perms = 0755

  backup = false
}
```
db.sh.ctmpl:
```
{{ range service "unique_tag.example_service@dc1" }}
  echo "{{ .Node }}"
{{ end }}
```

And ran (my host is named "Evariste":

```
consul-template -config db.conf
2018/11/19 16:44:02.212455 [WARN] vault.token: failed to renew: Put https://127.0.0.1:8200/v1/auth/token/renew-self: dial tcp 127.0.0.1:8200: connect: connection refused
2018/11/19 16:44:02.212480 [WARN] vault.token: renewer returned (maybe the lease expired)
Evariste.local
```

**Comment by bhn26** [comment]
@banks Here are some answers to your question:

Can we ask how you are registering the test service, I don't think I see that in the (great) description/repro steps:

We have all our nodes/services connect via their own VM with their own agent. The command that is ran looks like so:

```
/usr/bin/consul agent -config-file=/etc/consul.json -config-dir=/etc/consul.d -data-dir=/var/consul
```

The consul.json contains:
```
{
  "start_join": [{{ List of Consul servers to connect to }}],
  "disable_remote_exec": true,
  "disable_update_check": true,
  "data_dir": "/var/consul",
  "log_level": "WARN",
  "enable_syslog": true,
  "node_name": "test-client2",
  "server": false,
  "client_addr": "127.0.0.1",
  "bind_addr": "0.0.0.0",
  "advertise_addr": "1.2.3.4",
  "datacenter": "dc1",
  "encrypt": "{{ encrypt_key }}",
  "cert_file": "{{ Path to file }}",
  "key_file": "{{ Path to file }}",
  "ca_file": "{{ Path to file }}",
  "verify_outgoing": true,
  "ports": {
    "dns": 8600,
    "http": 8500,
    "https": 8501,
    "serf_lan": 8301,
    "serf_wan": 8302,
    "server": 8300
  },
  "enable_script_checks": true,
  "rejoin_after_leave": false,
  "leave_on_terminate": true
}
```

We have a json file in the consul.d/ directory defined like so:
```
{"service": {"name": "test-client", "id": "263183", "tags": [ "not_unique_tag", "unique_tag", "unique_tag2"], "check": {"name": "ttl-check", "DeregisterCriticalServiceAfter": "60s", "notes": "ttl-check", "ttl": "10s", "status": "passing"}}}
```

Can you also paste the output of the following against 1.2.3 and 1.3.0 please?
Output can be found here: https://gist.github.com/bhn26/878518c72c024562bd32972dd6d66de3

As with the configurations above, I've changed/omitted some variables but everything should be consistent with it all.

It's interesting because for 1.3.0 and 1.4.0, it returns an empty list. When I switch back to consul 1.2.3, it returns data as expected.

Finally, to confirm, are you running this against a single server or multiple servers? 

Multiple servers

Is CT talking directly to a server (yes if you are using dev mode) or to a local agent? 

Local agent

Is the local agent running the same Consul version as the server?

No, local agent is one 1.3.0 (At the time of testing) and the servers are on 1.2.1.

**Comment by bhn26** [comment]
I also forgot to mention that then I removed the "&tag=unique_tag" from:
```
curl -v 'localhost:8500/v1/health/service/example_service dc=dc1&stale=true&tag=unique_tag&wait=1m0s&pretty'
```

This returned data just fine (Just not as refined as desired).

**Comment by bhn26** [comment]
@banks Thanks for the input! I had always assumed it was both ways.

It was just an oversight during some of our testing and we noticed the behavior.
